### PR TITLE
[IMP] mail: limit followers rights for users with read access

### PR DESCRIPTION
**Pull request by ija-odoo**
PURPOSE

 Currently, if a user has read only  access to a record he can add
 or remove another follower to the channel, he can edit the subscription
 of other followers as both edit and remove icons for followers and
 'Add Followers' buttons are available for User.

So we need to restrict a user with read only access from add/remove/edit
other followers of the record.
  
SPECIFICATION

We resolved this by checking the access rights for Write permission of the
record, if the user has read only access (no write permission) then we will 
restrict users with a read only access to add/remove/edit the other followers
of this record, we aims at providing a middle ground between all or nothing,
thus he can only view other followers, edit his own subscription and
cannot remove, add or edit subscription of other followers  as both
edit and remove icons for followers and  'Add Followers' buttons are 
not available for user with read only access.

LINKS
PR #70575
Task 2447776

### CODE REVIEW

File: addons/mail/static/src/models/thread/thread.js
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:1888]: I would suggest to name it `hasWriteAccess` which is actually what it contains, instead of `canAddFollowers` which is just a consequence.
(ideally I would like to also have the idea that it's "for current user" in the field name but I don't have a good suggestion for that part that wouldn't be too long)

Also make sure to sort it alphabetically with other fields.

File: addons/mail/static/src/models/thread/thread.js
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:668]: no need to repeat the variable when it's the same name

File: addons/mail/static/src/models/thread/thread.js
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:636]: no need to repeat the variable when it's the same name

File: addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js:168]: also alphabetical here

File: addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js:317]: explicitly set `is_editable` to `false` in the other case as it's the main goal of the test

File: addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js:324]: This shouldn't be needed if `this.data` are set up correctly.

You might need to use `createChatterContainerComponent` to properly fetch data and then click on followers (rather than using `createFollowerListMenuComponent`).

File: addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js:389]: all the same remarks as in the other test

**Review by seb-odoo — COMMENTED** [review]
Hello, thank you for the PR, here are few comments.

File: addons/mail/controllers/discuss.py
Comment (seb-odoo) [inline addons/mail/controllers/discuss.py:431]: It should be possible to avoid all of this diff by computing `isEditable` in js directly (and remove `is_editable` from here), as long as `hasWriteAccess` is returned.

File: addons/mail/static/src/components/follower_list_menu/follower_list_menu.xml
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/follower_list_menu.xml:15]: This is old code but while I see it, followers is no longer support for channels at all, so `thread.channel_type !== 'chat'` should be `thread.model !== 'channel'`

File: addons/mail/static/src/components/follower_list_menu/follower_list_menu.xml
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/follower_list_menu.xml:21]: Probably should move the separator into a separate condition above the parent `t-if` that includes both the condition on write but also channel_type/channel, because this seems incomplete.

File: addons/mail/static/src/models/thread/thread.js
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:636]: guideline: alphabetical

File: addons/mail/static/src/models/thread/thread.js
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:668]: Doesn't need to be inside `followersData` condition, just add it into `values` at the start.

File: addons/mail/static/src/models/thread/thread.js
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:1994]: For the doc: the field is not specifically related to followers, but to the rights to edit the thread in general. It would be ok to specify this includes adding followers, but it should mention in priority its general meaning.

File: addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js:370]: typo: user "has"

File: addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js:374]: could be removed according to comments in python

File: addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js:308]: should not be necessary, cf other comments

**Review by seb-odoo — COMMENTED** [review]
follow up comments

File: addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js:406]: guideline: no empty line between user interaction and related asserts

File: addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js:337]: guideline: no empty line between user interaction and related asserts

File: addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js:341]: guideline: closing `)` on a new line, with same indent level as the line with the open `(`

File: addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js:410]: guideline: closing `)` on a new line, with same indent level as the line with the open `(`

File: addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js
Comment (seb-odoo) [inline addons/mail/static/src/components/follower_list_menu/tests/follower_list_menu_tests.js:347]: guideline (comment applies in the whole diff): "logged in user" is typically called "current user" instead (because generally speaking another user could also be "logged in" on his own side)

File: addons/mail/static/src/models/follower/follower.js
Comment (seb-odoo) [inline addons/mail/static/src/models/follower/follower.js:139]: lint: space before `?`

File: addons/mail/static/src/models/follower/follower.js
Comment (seb-odoo) [inline addons/mail/static/src/models/follower/follower.js:158]: Missing dot at the end of the sentence.

File: addons/mail/static/src/models/thread/thread.js
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:648]: lint: space inside inline object `{ ... }`

**Review by seb-odoo — APPROVED** [review]
Thank you for the update, last few clean up comments, but otherwise it looks good to me.

I'll let @odoo/rd-sm have the final check/follow standard procedure as the task is in their project.

File: addons/mail/static/src/components/follower_list_menu/follower_list_menu.xml
Comment (dja-odoo) [reply addons/mail/static/src/components/follower_list_menu/follower_list_menu.xml:8]: @seb-odoo  Indeed that should be the case. We are hiding the dropdown only when there are no followers `AND` user does not has write access to the record (because there is nothing to show). Even if there's a single follower, the list will be opened. The test cases are also written / adapted for all the cases.

Comment (dja-odoo) [reply addons/mail/static/src/components/follower_list_menu/follower_list_menu.xml]: @seb-odoo Should be okay now.

**Comment by alexkuhn** [comment]
I've made some comments on commit:
https://github.com/odoo/odoo/commit/472fac659ed33302158be8dd0ab5c588ddbea3d2

I've created a Task as a result (Task-2803053), and I'll make the fix myself. Thanks!

**Pull request closed without merging**
